React/testing.md

```diff
@@ -1 +1,299 @@
+
+The default test in `App.test.js` first creates a `div` element which is **not** embedded in any webpage and then embeds an instance of the App component inside the `div`.  In this way it tests that the component can be rendered at all.  Lastly `ReactDOM.unmountComponentAtNode` destroys the instance to save memory.
+
+This test doesn't actually make any assertions, there are no `expect(div).toBe(y)` commands, but it ensure that the component will show up on the screen without crashing React.
+
+### Installing Enzyme & React Test Renderer
+
+Enzyme is a handy library which will make it easier to manipulate, traverse and make assertions on our React components.  It's made to work similar to jQuery for selecting, rendering and manipulating the DOM.
+
+To install enzyme for testing your application, execute from the terminal:
+
+```bash
+npm i --save-dev enzyme enzyme-adapter-react-16
+```
+
+**Note** This will install enzyme and an adapter or version 16 of React.  You can expect a new adapter for later versions of React.
+
+Next tell Jest how to configure enzyme and use the adapter.  Create the file `src/setupTests.js` and add the following content.
+
+```javascript
+// src/setupTests.js
+import { configure } from 'enzyme';
+import Adapter from 'enzyme-adapter-react-16';
+
+configure({ adapter: new Adapter() });
+```
+
+This file is run when you start testing with create-react-app and allows us to tell enzyme which adapter to use.  If you have tests running you will need to exit and restart Jest.
+
+## Snapshot Testing
+
+### What is it?
+
+With React we often want to ensure that a component is rendering properly.  For our `NewPetForm` component we could write tests ensure that all the fields are rendering through tests that:
+
+1.  Ensure that the component renders 4 input fields.
+2.  Each field is labeled properly.
+3.  Ensure there is a submit button
+4.  and more and more and more...
+
+However these tests are pretty fragile, easy to break when you make minor changes to the form and tedius to update for UI refactoring.
+
+Instead we can:
+1.  Convert the rendered html of the component into JSON with the JSON holding the most significant features of the component.
+2.  Save that JSON to a "Snapshot" file.
+3.  Compare future runs of the test to that snapshot and alert the user to any changes.
+
+When a component changes, we can be alerted and make corrections, or update the snapshot for future test runs.
+
+**Question:** Can you do TDD with a Snapshot test?
+
+### Creating our First Snapshot
+
+Lets create a snapshot test for one of our untested components, `NewPetForm`.
+
+First create `src/NewPetForm.test.js`
+
+```javascript
+// src/NewPetForm.test.js
+import React from 'react';
+import NewPetForm from './NewPetForm';
+import { mount } from 'enzyme';
+
+describe('NewPetForm', () => {
+  test('that it matches an existing snapshot', () => {
+    const wrapper = mount( <NewPetForm addPetCallback={() => {} } />);
```

Comment:
I'd love to see a line-by-line explanation of what all of this code is doing  in the test block - maybe comments on each line?